### PR TITLE
Add build nextflow schemas action

### PR DESCRIPTION
**Pull request by DriesSchaumont**
Requires https://github.com/viash-io/viash-actions/pull/7 to be merged first, and after the branches in this PR should be adjusted

### CODE REVIEW

File: build_nextflow_schemas/action.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        viash ns list -s ${{ inputs.input_dir }} --format json 2> /dev/null > ${{ runner.temp }}/ns_list_src.json
+        [ -s ${{ runner.temp }}/ns_list_src.json ] || { echo >&2 "No workflow Viash config files found."; exit 1; }
+        inputs=$(jq -r '[.[] | .info.config] | join(";")' ${{ runner.temp }}/ns_list_src.json)
+        [ -z "$inputs" ] && { echo >&2 "No workflow Viash config files found."; exit 1; }
+        if [ ! -z "${{inputs.output_directory}}" ]; then
+          # Default behavior
+          jq_query='[.[] | "'"${{inputs.output_directory}}"'/nextflow/" + .functionality.namespace + "/" + .functionality.name + "/nextflow_schema.json"] | join(";")'
+        else
+          jq_query='[.[] | .info.config | capture("^(?<dir>.*\/)").dir + "/nextflow_params.yaml"] | join(";")'
+        fi
```
Comment (rcannood) [inline build_nextflow_schemas/action.yml:68]: This switch is a bit awkward but I don't know how else I would solve it at this stage.

:thinking: :thinking: 

How about this? We define a `output_pattern` parameter and let the default be `{dir}/nexflow_schema.json`. You can also set it to `target/nextflow/{namespace}/{functionality-name}/nextflow_schema.json` to get the alternative behaviour.

You can then use the following to get `outputs_schema`:

```bash
exec_query="target/nextflow/{namespace}/{functionality-name}/nextflow_schema.json"
outputs_schema=$(viash ns exec "echo $exec_query;" 2> /dev/null | tr -d '\n')
```

**Review by rcannood — CHANGES_REQUESTED** [review]
* Could you add something to the `/README.md` and add a `build_nextflow_schemas/README.md`?
* Could you rename the action to `build-nextflow-schemas` in line with `viash-actions` and `ns-list`?

See other comments as well, especially the `jq_query` one!

File: build_nextflow_schemas/action.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        inputs=$(jq -r '[.[] | .info.config] | join(";")' << 'EOF'
+        ${{ steps.nslist.outputs.output }}
+        EOF
+        )
+        [ -z "$inputs" ] && { echo >&2 "No workflow Viash config files found."; exit 1; }
+        if [ ! -z "${{inputs.output_directory}}" ]; then
+          # Default behavior
+          jq_query='[.[] | "'"${{inputs.output_directory}}"'/nextflow/" + .functionality.namespace + "/" + .functionality.name + "/nextflow_schema.json"] | join(";")'
+        else
+          jq_query='[.[] | .info.config | capture("^(?<dir>.*\/)").dir + "/nextflow_params.yaml"] | join(";")'
```
Comment (rcannood) [inline build_nextflow_schemas/action.yml:75]: Watch out, this should be `nextflow_schema.json`.

File: build_nextflow_schemas/action.yml (hunk 1 above)
Comment (rcannood) [reply build_nextflow_schemas/action.yml:76]: ### Alternative

Or maybe we can make it work like this:

```yaml
    - uses: viash-io/build-nextflow_schemas
      with:
        # outputs to `target/nextflow/{namespace}/{name}/nextflow_schema.json
        # uses `-p nextflow` to make sure the component has a nextflow platform
        components: src/
        # outputs to `{src-dir}/nextflow_schema.json`.
        # should only run when `type: nextflow_script` is defined in the resources
        workflows: workflows/
```

### Yet another alternative

_Or_, this is also fine by me:

```yaml
    - uses: viash-io/build-nextflow_schemas
      with:
        src: src/
        type: component
    - uses: viash-io/build-nextflow_schemas
      with:
        src: workflows/
        type: workflows
```

File: build_nextflow_schemas/action.yml (hunk 2 above)
Comment (DriesSchaumont) [reply build_nextflow_schemas/action.yml]: Fixed!

File: build_nextflow_schemas/action.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        ref: ${{ steps.latestrelease.outputs.releasetag }}
+    - name: 'Generate schemas'
+      shell: bash
+      id: generate-schemas
+      run: |
+        if ! command -v viash &> /dev/null
+        then
+          echo "Viash could not be found.";
+          exit 127;
+        fi
```
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply build_nextflow_schemas/action.yml]: Good suggestion, I propose that we tackle this later in a separate PR

File: build_nextflow_schemas/action.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        viash ns list -s ${{ inputs.input_dir }} --format json 2> /dev/null > ${{ runner.temp }}/ns_list_src.json
+        [ -s ${{ runner.temp }}/ns_list_src.json ] || { echo >&2 "No workflow Viash config files found."; exit 1; }
```
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply build_nextflow_schemas/action.yml]: Done!

File: build_nextflow_schemas/action.yml (hunk 1 above)
Comment (DriesSchaumont) [reply build_nextflow_schemas/action.yml]: Done, used the following solution:

```yaml
    - uses: viash-io/build-nextflow_schemas
      with:
        # outputs to `target/nextflow/{namespace}/{name}/nextflow_schema.json
        # uses `-p nextflow` to make sure the component has a nextflow platform
        components: src/
        # outputs to `{src-dir}/nextflow_schema.json`.
        # should only run when `type: nextflow_script` is defined in the resources
        workflows: workflows/
```

File: build_nextflow_schemas/action.yml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+          component_outputs_schema=$(jq -r "$jq_query" <<-'  EOF'
+          ${{ steps.nslistcomponents.outputs.output }}
+          EOF
+          )
+        fi
+
+        # Process workflows
+        if [ "${{ steps.nslistworkflows.outputs.output != null }}" == "true" ]; then
+          # Get config paths
+          workflow_inputs=$(jq -r '[.[] | .info.config] | join(";")' <<-'  EOF'
```
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply build_nextflow_schemas/action.yml]: Forgot about that. Added!

File: build_nextflow_schemas/README.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+```yaml
+name: demo of generate_documentation_qmd
+on:
+  pull_request:
+jobs:
+  demo:
+    runs-on: ubuntu-latest
+    
+    steps:
+    - uses: viash-io/viash-actions/setup@v1.0.0
```
Comment (rcannood) [inline build_nextflow_schemas/README.md:36]: switch to v2

**Review by rcannood — APPROVED** [review]

**Pull request merged**